specs/GetFavicon.md

```diff
@@ -0,0 +1,172 @@
+# Background
+A favicon (favorite icon) is a tiny icon included along with a website, which is displayed in places like the browser's address bar, page tabs and bookmarks menu. Developers would like to have an API which allows them to retrieve the Favicon of a webpage, if it has been set, as well as get an update whenever the favicon has changed.
+
+# Description
+We propose a new Webview2 event which would allow developers to access the current Favicon of a page and be notified when the favicon changes. This means when a page first loads, it would raise the FaviconChanged event since before parsing the HTML document there is no known favicon. DOM or JavaScript may change the Favicon, causing the event to be raised again for the same document.
+
+# Examples
+## Win32 C++ Registering a listener for favicon changes
+```cpp
+Gdiplus::GdiplusStartupInput gdiplusStartupInput;
+
+// Initialize GDI+.
+Gdiplus::GdiplusStartup(&gdiplusToken_, &gdiplusStartupInput, NULL);
+CHECK_FAILURE(m_webView2->add_FaviconChanged(
+    Callback<ICoreWebView2ExperimentalFaviconChangedEventHandler>(
+        [this](ICoreWebView2* sender, IUnknown* args) -> HRESULT {
+            if (m_faviconChanged)
+            {
+                wil::unique_cotaskmem_string url;
+                Microsoft::WRL::ComPtr<ICoreWebView2>
+                    webview2Experimental;
+                CHECK_FAILURE(
+                    sender->QueryInterface(IID_PPV_ARGS(&webview2)));
+
+                CHECK_FAILURE(webview2Experimental->get_FaviconUri(&url));
+                std::wstring strUrl(url.get());
+
+                webview2->GetFavicon(
+                    COREWEBVIEW2_FAVICON_IMAGE_FORMAT_PNG,
+                    Callback<ICoreWebView2ExperimentalGetFaviconCompletedHandler>(
+                        [this, strUrl](HRESULT errorCode, IStream* iconStream) -> HRESULT
+                        {
+                            CHECK_FAILURE(errorCode);
+                            Gdiplus::Bitmap iconBitmap(iconStream);
+                            wil::unique_hicon icon;
+                            if (!iconBitmap.GetHICON(&icon))
```

Comment:
Is this GetHICON check the place where sample code is noticing the favicon is empty (e.g. unspecified)?  (I don't remember offhand if that "fails" if the iconStream was empty)

Reply:
The failure is first detected at `Gdiplus::Bitmap iconBitmap(iconStream);` - the `iconBitmap` is in an error state: If we were to call `GetLastStatus()`, it would return `InvalidParameter`. We ignore that error and just call `GetHICON` anyway, which also fails with `InvalidParameter`.

Reply:
I have validated and added some explicit code to how to handle "empty icons".